supervisor/docker/progress.py

```diff
@@ -0,0 +1,132 @@
+"""Docker pull progress emitter."""
+import asyncio
+import time
+from contextlib import suppress
+
+import docker
+
+from ..coresys import CoreSysAttributes
+from ..exceptions import HomeAssistantAPIError
+
+
+def pull_with_progress(interface: CoreSysAttributes, image, tag):
+    """Docker pull with progress"""
+
+    progress = PullProgress(interface)
+    try:
+        progress.send_event()
+        pull_log = interface.sys_docker.api.pull(image, tag, stream=True, decode=True)
+        progress.process_log(pull_log)
+
+        return interface.sys_docker.images.get(
+            "{0}{2}{1}".format(image, tag, "@" if tag.startswith("sha256:") else ":")
+        )
+    except docker.errors.APIError as err:
+        progress.done()
+        raise err
+
+
+class Status:
+    """Docker image status object"""
+
+    def __init__(self):
+        self._current = {}
+        self._total = {}
+
+    def update(self, layer_id, current, total):
+        """Update one layer status"""
+        self._current[layer_id] = current
+        self._total[layer_id] = total
+
+    def done(self, layer_id):
+        """Mark one layer as done"""
+        if layer_id in self._total:
+            self._current[layer_id] = self._total[layer_id]
+
+    def done_all(self):
+        """Mark image as done"""
+        if len(self._total) == 0:
+            self.update("id", 1, 1)
+        for layer_id in self._total:
+            self._current[layer_id] = self._total[layer_id]
+
+    def current(self):
+        """Get current for image"""
+        return sum(self._current.values())
+
+    def total(self):
+        """Get total for image"""
+        return sum(self._total.values())
+
+
+class PullProgress:
+    """Docker pull log progress listener"""
+
+    def __init__(self, interface: CoreSysAttributes, sleep=1.0) -> None:
+        super().__init__()
+        self._interface = interface
+        self._sleep = sleep
+        self._next_push = 0
+        self._download = Status()
+        self._extract = Status()
+
+    def send_event(self):
+        """Send event to HA Core"""
+        self._next_push = time.time() + self._sleep
+        asyncio.run_coroutine_threadsafe(
+            self._send_event(self._status()), self._interface.sys_loop,
+        )
+
+    def done(self):
+        """Mark current pull as done and send this info to HA Core"""
+        self._download.done_all()
+        self._extract.done_all()
+        self.send_event()
+
+    async def _send_event(self, status) -> None:
+        with suppress(HomeAssistantAPIError):
+            async with self._interface.sys_homeassistant.make_request(
+                "post", "api/events/hassio_progress", json=status, timeout=2,
```

Comment:
We are not allow to use the event/service bus of the core. I would suggest to make an `/{object}/progress` endpoint which can be called by UI. If they return `None` there is no job running, otherwise we return information to show the progressbar.

Reply:
Do you mean we should poll this endpoint?
I think it is not the best way.
What do you think about creating a new WebSocket connection from the UI to the supervisor?

Reply:
If you want stream the logs, we need change the library to aiodocker or other asyncio library, otherwise we need change the socket object into nonblock and wrap it with a async streamreader/writter.

However, what you want do is a ton of work over multible repository, 1-3 month. Feel free to give a shot, but maybe make first an RFC to reduce the changes with reviews.

Reply:
Pascal and I discussed this offline. We should do both.

We want an event to indicate that something has changed, and then an API endpoint that gives us the details. We shouldn't include all the data in the event.

Reply:
Do you mean an event in the core, as I did?
What would be the next step? Help me a little :)

Reply:
I need write an RFC for an job based API based on `process_lock` and the new contextvars which write/manage jobs global and can send events to the bus. But I'm also in vacation for next weeks and limited what I can do.

Reply:
@Sanyatuning next step will be to wait until Pascal is back from parental leave and design a more elaborate job-progress system. We don't want to hack it into just the Docker API as there might be other tasks that we want to report on (backups etc).

Reply:
Thanks for the explanation.